src/backend/dx12/src/command.rs

```diff
@@ -602,6 +616,182 @@ impl CommandBuffer {
+                        },
+                        buf_offset: image::Offset {
+                            x: 0,
+                            .. buf_offset
+                        },
+                        copy_extent: image::Extent {
+                            width: r.image_extent.width - half,
+                            .. r.image_extent
+                        },
+                    });
+                }
+            } else {
+                // worst case: row by row copy
+                assert_eq!(image.block_dim, (1, 1)); // TODO
+                for z in 0 .. r.image_extent.depth {
+                    for y in 0 .. r.image_extent.height {
+                        // an image row starts non-aligned
+                        let row_offset = layer_offset +
+                            z as u64 * slice_pitch as u64 +
+                            y as u64 * row_pitch as u64;
+                        let aligned_offset = row_offset & !(d3d12::D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT as u64 - 1);
+                        let next_aligned_offset = aligned_offset + d3d12::D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT as u64;
+                        let cut_width = cmp::min(r.image_extent.width, (next_aligned_offset - row_offset) as _);
+                        let gap = (row_offset - aligned_offset) as u32;
+
+                        copies.push(Copy {
+                            footprint_offset: aligned_offset,
+                            footprint: image::Extent {
+                                width: cut_width + gap,
+                                height: 1,
+                                depth: 1,
+                            },
+                            row_pitch: d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT,
+                            img_subresource,
+                            img_offset: image::Offset {
+                                x: r.image_offset.x,
+                                y: r.image_offset.y + y as i32,
+                                z: r.image_offset.z + z as i32,
+                            },
+                            buf_offset: image::Offset {
+                                x: gap as i32,
+                                y: 0,
+                                z: 0,
+                            },
+                            copy_extent: image::Extent {
+                                width: cut_width,
+                                height: 1,
+                                depth: 1,
+                            },
+                        });
+
+                        // and if it crosses a pitch alignment - we copy the rest separately
+                        if cut_width == r.image_extent.width {
+                            continue;
+                        }
+                        let leftover = r.image_extent.width - cut_width;
+
+                        copies.push(Copy {
+                            footprint_offset: next_aligned_offset,
```

Comment:
This looks somewhat suspicious to me as `D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT` is twice as large as the required row pitch.